### PR TITLE
Remove Throwable from Ref

### PR DESCRIPTION
**Pull request by SystemFw**
As per the title, it partially addresses #993 .
 Note that the behaviour of all code that depends on Ref within fs2 has _not_ been changed, e.g. I haven't experimented with changing on what EC anything runs, or switching async to sync or anything like that.
This is a sheer refactor, that shouldn't affect the behaviour of the rest of fs2 at all. The semantics of Ref have slightly changed though (hopefully for the best), and the implementation should be a bit clearer without Throwable. 
The commits are self-contained if you prefer to look at them them individually.
Let me know what you think.

### CODE REVIEW

File: core/shared/src/main/scala/fs2/async/Ref.scala [hunk 1]
```diff
@@ -150,9 +153,13 @@ final class Ref[F[_],A] private[fs2] (implicit F: Effect[F], ec: ExecutionContex
       // or the actor directly, and the winner destroys any
       // references behind it!
       if (won.compareAndSet(false, true)) {
-        val actor = ref.get
-        ref.set(null)
-        actor ! Msg.Set(res, () => ())
+        res match {
+          case Left(e) => throw e
```
Comment (mpilquist) [inline core/shared/src/main/scala/fs2/async/Ref.scala]: Should the `ref.set(null)` still happen in this case?

Comment (SystemFw) [reply core/shared/src/main/scala/fs2/async/Ref.scala]: Good catch! I'll have to think about it

Comment (SystemFw) [reply core/shared/src/main/scala/fs2/async/Ref.scala]: Probably yes, or the loser still has a reference, and won't be garbage collected when the winner fails

**Comment by mpilquist** [comment]
LGTM. Not sure about the one TODO listed and the one comment I left. Otherwise looks ready to merge to me.

**Comment by SystemFw** [comment]
Yeah, I wanted to check the general approach was OK. I think your comment is spot on, and I'll have to think a tad more about that TODO item, but I'll probably be able to finish today

**Comment by SystemFw** [comment]
Wrt the Boolean in the callback to the Nevermind message to the actor, it's used to signal that the `id` of the listener you want to cancel doesn't correspond to any listener. However, the only place that sends a Nevermind message is `cancellableGet`, which however is sure that the `id` exists (since it's the one of the listener it just created), and therefore ignores the Boolean. I'm inclined to remove it.

**Pull request merged**